bitcoin/src/pow.rs

```diff
@@ -0,0 +1,1530 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bitcoin::Work;
+    /// let w = Work::from(0xdeadbeef_u32);
+    /// assert_eq!(w.to_target().to_work(), w)
+    /// ```
+    pub fn to_target(self) -> Target { Target(self.0.inverse()) }
+}
+do_impl!(Work);
+
+impl Add for Work {
+    type Output = Work;
+    fn add(self, rhs: Self) -> Self { Work(self.0.wrapping_add(rhs.0)) }
+}
+
+impl Sub for Work {
+    type Output = Work;
+    fn sub(self, rhs: Self) -> Self { Work(self.0.wrapping_sub(rhs.0)) }
+}
+
+/// A 256 bit integer representing target.
+///
+/// The SHA-256 hash of a block's header must be lower than or equal to the current target for the
+/// block to be accepted by the network. The lower the target, the more difficult it is to generate
+/// a block. (See also [`Work`].)
+///
+/// ref: <https://en.bitcoin.it/wiki/Target>
+///
+/// # Examples
+///
+/// ```
+/// use bitcoin::Target;
+/// let _ = Target::from(0xdeadbeef_u32);
+/// ```
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Target(U256);
+
+impl Target {
+    /// The maximum possible target.
+    ///
+    /// This value is used to calculate difficulty, which is defined as how difficult the current
+    /// target makes it to find a block relative to how difficult it would be at the highest
+    /// possible target. Remember highest target == lowest difficulty.
+    ///
+    /// ref: <https://en.bitcoin.it/wiki/Target>
+    // In Bitcoind this is ~(u256)0 >> 32 stored as a floating-point type so it gets truncated, hence
+    // the low 208 bits are all zero.
+    const MAX: Self = Target(U256(0xFFFF_u128 << (208 - 128), 0));
+
+    /// The maximum possible target (see [`Target::MAX`]).
+    ///
+    /// This is provided for compatibility with Rust 1.41.1, newer code should use [`Target::MAX`].
+    pub fn max_value() -> Self { Target::MAX }
+
+    /// Computes the [`Target`] value from a compact representation.
+    ///
+    /// ref: <https://developer.bitcoin.org/reference/block_chain.html#target-nbits>
+    pub fn from_compact(c: CompactTarget) -> Target {
+        let bits = c.0;
+        // This is a floating-point "compact" encoding originally used by
+        // OpenSSL, which satoshi put into consensus code, so we're stuck
+        // with it. The exponent needs to have 3 subtracted from it, hence
+        // this goofy decoding code. 3 is due to 3 bytes in the mantissa.
+        let (mant, expt) = {
+            let unshifted_expt = bits >> 24;
+            if unshifted_expt <= 3 {
+                ((bits & 0xFFFFFF) >> (8 * (3 - unshifted_expt as usize)), 0)
+            } else {
+                (bits & 0xFFFFFF, 8 * ((bits >> 24) - 3))
+            }
+        };
+
+        // The mantissa is signed but may not be negative.
+        if mant > 0x7F_FFFF {
+            Target::default()
```

Comment:
I find this confusing. What is a default target? We should have a meaningful constant, not the trait.

Reply:
I added private const `Target::ZERO`.